src/main/java/org/apache/commons/crypto/stream/CryptoOutputStream.java

```diff
@@ -298,30 +301,47 @@ public boolean isOpen() {
     @Override
     public int write(ByteBuffer src) throws IOException {
         checkStream();
-        final int len = src.remaining();
-        int remaining = len;
+
+        if (transferInProgress && outBuffer.hasRemaining()) {
+            output.write(outBuffer);
+            return 0;
+        } else {
+            outBuffer.clear();
+            transferInProgress = false;
+        }
+
+        int bytesWritten = 0;
+        int remaining = src.remaining();
+
         while (remaining > 0) {
-            final int space = inBuffer.remaining();
-            if (remaining < space) {
+            int written;
+            int space = inBuffer.remaining();
+
+            if(remaining < space) {
                 inBuffer.put(src);
-                remaining = 0;
+                bytesWritten += remaining;
+                break;
```

Comment:
Hmm... here you're just buffering data without even trying to write to the output. Is that intentional? Feels like you should `encrypt()` and try to write stuff even if all the input buffer fits in the internal buffer.

But really there might be a disconnect here that I don't know is possible to fix. If the input buffer has `x` bytes, and they fit in the internal buffer, but you can only write `y` bytes (`y < x`) to the underlying channel, what happens?

If you return "I wrote x bytes", the caller may expect that those bytes will be in the final output, when that might not be the case. If you return "I wrote y bytes", it assumes that y bytes of plain text are equal to y bytes of cipher text. If you return 0, you may mess up the caller (since you've actually consumed input but are reporting 0 bytes written).

I'm starting to think that it might be better to leave this as a blocking interface, and leave it to applications to deal with non-blocking channels. e.g., thinking about Spark, you can't hook this `WritableByteChannel` implementation cleanly. It actually makes more sense to let Spark handle the data encryption, much like this code is doing internally, and do its own bookkeeping of what has been written and what has been not.